src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,206 @@
+/// to value for each range.
+pub struct RangeAggregation {
+    /// The field to aggregate on.
+    pub field_name: String,
+    /// Note that this aggregation includes the from value and excludes the to value for each
+    /// range. Extra buckets will be created until the first to, and last from.
+    pub buckets: Vec<Range<f64>>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeBucketEntry {
+    range: Range<u64>,
+    bucket: SegmentBucketDataEntry,
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentRangeBucketEntry>,
+    field_type: Type,
+}
+
+impl SegmentRangeCollector {
+    pub fn into_intermediate_agg_result(self) -> IntermediateBucketResult {
+        let field_type = self.field_type.clone();
+        let buckets = self
+            .buckets
+            .into_iter()
+            .map(move |range_bucket| {
+                (
+                    range_to_key(&range_bucket.range, &field_type),
+                    range_bucket.bucket.into(),
+                )
+            })
+            .collect();
+
+        IntermediateBucketResult { buckets }
+    }
+
+    pub(crate) fn from_req(
+        req: &RangeAggregation,
+        sub_aggregation: &AggregationsWithAccessor,
+        field_type: Type,
+    ) -> Self {
+        // The range input on the request is f64.
+        // We need to convert to u64 ranges, because we read the values as u64.
+        // The mapping from the conversion is monotonic so ordering is preserved.
+        let buckets = extend_range(&req.buckets, &field_type)
+            .iter()
+            .map(|range| SegmentRangeBucketEntry {
+                range: range.clone(),
+                bucket: SegmentBucketDataEntry::KeyCount(SegmentBucketDataEntryKeyCount {
+                    key: range_to_key(&range, &field_type),
+                    doc_count: 0,
+                    values: None,
+                    sub_aggregation: SegmentAggregationResultsCollector::from_req(sub_aggregation),
+                }),
+            })
+            .collect();
+
+        SegmentRangeCollector {
+            buckets,
+            field_type,
+        }
+    }
+
+    #[inline]
+    pub(crate) fn collect(
+        &mut self,
+        doc: DocId,
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+    ) {
+        let val = bucket_with_accessor.accessor.get(doc);
+        self.collect_val(val, doc, bucket_with_accessor);
+    }
+
+    #[inline]
+    fn collect_val(
+        &mut self,
+        val: u64,
+        doc: DocId,
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+    ) {
+        let bucket = self
+            .buckets
+            .iter_mut()
+            .find(|bucket| bucket.range.contains(&val))
```

Comment:
:-|

Ok so this is really terrible code here.
For one, I don't think the partition covers all of the space...

Then elasticsearch weirdly decided to allow for overlapping ranges.
Finally, if we do not accept overlapping buckets, then a binary search would have been way more efficient.

I suggest:
- We refuse overlapping buckets.
- We artificially extend the partition to cover the entire space. That means we add two buckets starting from u64::min and ending at u64::max collector to cover everything.
- We represent the bucket scale using a Vec<u64>
- We remap the values to the correct buckets.
- We use a binary search.

If we have a need for overlapping buckets in the future, we can add a specific/slower implementation for them.

Reply:
binary search is not necessarily much more efficient because of the low cache efficiency. It depends on the number of buckets and if the incoming data is sorted (improved branch prediction).

11 buckets, data sorted
```
test aggregation::bucket::range::bench::bench_small_range_contains_binary_search                                         ... bench:     126,452 ns/iter (+/- 9,607)
test aggregation::bucket::range::bench::bench_small_range_contains_linear_search                                         ... bench:     163,070 ns/iter (+/- 24,590)
```

5 buckets, data sorted
```
test aggregation::bucket::range::bench::bench_small_range_contains_binary_search                                         ... bench:      83,141 ns/iter (+/- 3,984)
test aggregation::bucket::range::bench::bench_small_range_contains_linear_search                                         ... bench:      81,279 ns/iter (+/- 8,091)
```

> If we have a need for overlapping buckets in the future, we can add a specific/slower implementation for them.

Yes, that should be easy to add later. Performance wise we can expect a little more than 2x

Reply:
5 and 11 buckets is very small. Think about a time histogram.